// cape-lodash/lib/dots.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.renamePick = exports.rename = exports.fpMove = exports.fpCopy = undefined;

var _unset2 = require('lodash/unset');

var _unset3 = _interopRequireDefault(_unset2);

var _set2 = require('lodash/set');

var _set3 = _interopRequireDefault(_set2);

var _reduce2 = require('lodash/reduce');

var _reduce3 = _interopRequireDefault(_reduce2);

var _rearg2 = require('lodash/rearg');

var _rearg3 = _interopRequireDefault(_rearg2);

var _get2 = require('lodash/get');

var _get3 = _interopRequireDefault(_get2);

var _flip2 = require('lodash/flip');

var _flip3 = _interopRequireDefault(_flip2);

var _curry2 = require('lodash/curry');

var _curry3 = _interopRequireDefault(_curry2);

var _ary2 = require('lodash/ary');

var _ary3 = _interopRequireDefault(_ary2);

exports.copy = copy;
exports.move = move;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function copy(getKey, setKey, source, target) {
  var value = (0, _get3.default)(source, getKey);
  return value === undefined ? target : (0, _set3.default)(target, setKey, value);
}
var fpCopy = exports.fpCopy = (0, _curry3.default)((0, _rearg3.default)(copy, [3, 2, 0, 1]), 4);
function move(getKey, setKey, object) {
  (0, _set3.default)(object, setKey, (0, _get3.default)(object, getKey));
  (0, _unset3.default)(object, getKey);
  return object;
}
var fpMove = exports.fpMove = (0, _curry3.default)((0, _ary3.default)((0, _flip3.default)(move), 3));
var rename = exports.rename = (0, _curry3.default)(function (renameObj, source) {
  return (0, _reduce3.default)(renameObj, fpMove, source);
});
// key is get, value is set.
var renamePick = exports.renamePick = (0, _curry3.default)(function (renameObj, source) {
  return (0, _reduce3.default)(renameObj, fpCopy(source), {});
});
