// cape-lodash/lib/set.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.replaceField = exports.setFieldHas = exports.setField = exports.setIn = exports.setVal = exports.setKeyVal = exports.setKey = undefined;

var _has2 = require('lodash/fp/has');

var _has3 = _interopRequireDefault(_has2);

var _rearg2 = require('lodash/rearg');

var _rearg3 = _interopRequireDefault(_rearg2);

var _get2 = require('lodash/get');

var _get3 = _interopRequireDefault(_get2);

var _curry2 = require('lodash/curry');

var _curry3 = _interopRequireDefault(_curry2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.set = set;

var _cond = require('./cond');

var _transform = require('./transform');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toArray(arr) { return Array.isArray(arr) ? arr : Array.from(arr); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function set(state, key, value) {
  return _extends({}, state, _defineProperty({}, key, value));
}
var setKey = exports.setKey = (0, _curry3.default)((0, _rearg3.default)(set, [1, 0, 2]), 3);
var setKeyVal = exports.setKeyVal = (0, _curry3.default)((0, _rearg3.default)(set, [2, 0, 1]), 3);
var setVal = exports.setVal = (0, _curry3.default)((0, _rearg3.default)(set, [1, 2, 0]), 3);
var setIn = exports.setIn = (0, _curry3.default)(function (_ref, state, value) {
  var _ref2 = _toArray(_ref),
      key = _ref2[0],
      rest = _ref2.slice(1);

  if (!rest.length) return set(state, key, value);
  return set(state, key, setIn(rest, (0, _get3.default)(state, key, {}), value));
});
var setField = exports.setField = (0, _curry3.default)(function (path, transformer) {
  return function (item) {
    return set(item, path, transformer(item));
  };
});
var setFieldHas = exports.setFieldHas = (0, _curry3.default)(function (path, transformer) {
  return (0, _cond.condId)([(0, _has3.default)(path), setField(path, transformer)]);
});
var replaceField = exports.replaceField = (0, _curry3.default)(function (path, transformer) {
  return (0, _cond.condId)([(0, _has3.default)(path), setField(path, (0, _transform.doProp)(transformer, path))]);
});
